lib/SILGen/SILGenDestructor.cpp

```diff
@@ -497,113 +497,59 @@ void SILGenFunction::emitClassMemberDestruction(ManagedValue selfValue,
 
 void SILGenFunction::emitMoveOnlyMemberDestruction(SILValue selfValue,
```

Comment:
Between `ownershipmodeleliminatorvisitor::splitDestroy` and `silgenfunction::emitMoveOnlyMemberDestruction` I'm not really sure who's suppose to be expanding the destruction now.

Seems like the intention is that we emit this in SILGen:
```
%dd = drop_deinit self
destroy_value %dd
```

and then when we lower out of OSSA we expand that `destroy(drop_deinit)` like we use to in SILGen with memberwise-destruction, like:

```
%elms = destructure_struct self
destroy %elm1
...
```

This model seems nicer; it leaves the decision of how to destroy the value up to SIL, and uses just the marker `drop_deinit` to know that a release is not how to do it since it triggers deinit.

Reply:
@kavon Right. Unfortunately `drop_deinit` can also be used on address types, in which case it means something significantly different. The interesting features of `drop_deinit` are the OSSA/lifetime semantics. The address form doesn't have any of those semantics. It's just a placeholder to indicate that we need to be careful when promoting the address to OSSA. We should to add a separate `drop_deinit_addr`, but I didn't get to that in this PR.

A couple things to note:

SILGen *could* legally expand the `destroy_value` into a `destructure_struct`, but that would be silly and non-canonical.

The optimizer can promote an address `drop_deinit` to a value (mem2reg), but when doing so, we need to check the same conditions present in the SILVerifier... it can only be used to destroy or destructure.